File: tdm/domain/progress.py
# src-pyloid/progress.py
from __future__ import annotations
from typing import Callable, Literal, Optional
import time

Level = Literal["info", "success", "warning", "error"]
Status = Literal["running", "done", "error"]

class Progress:
    """
    .info/.warning/.error 등으로 호출.
    emit(payload: dict) 콜백으로 외부(RPC progress dict / 이벤트 버스)로 전달.
    """
    def __init__(self, emit: Callable[[dict], None], total: Optional[int] = None):
        self.emit_cb = emit
        self.step_no = 0
        self.total = total
        self.phase_step: Optional[int] = None
        self.phase_total: Optional[int] = None

    def _post(
        self,
        message: str,
        level: Level = "info",
        status: Status = "running",
        inc: bool = False,
        *,
        error: Optional[str] = None,
        detail: Optional[str] = None,
    ):
        if inc:
            self.step_no += 1
        payload = {
            "ts": time.time(),
            "step": self.step_no,
            "total": self.total,
            "phase_step": self.phase_step,
            "phase_total": self.phase_total,
            "level": level,
            "status": status,
            "message": message,
        }
        if error is not None:
            payload["error"] = error
        if detail is not None:
            payload["detail"] = detail
        self.emit_cb(payload)

    def info(self, msg: str, *, inc: bool = False):    self._post(msg, "info",    "running", inc)
    def success(self, msg: str, *, inc: bool = False): self._post(msg, "success", "running", inc)
    def warning(self, msg: str, *, inc: bool = False): self._post(msg, "warning", "running", inc)
    def error(self, msg: str, *, detail: Optional[str] = None, inc: bool = False):
        self._post(msg, "error", "error", inc, error=msg, detail=detail)
    def step(self, msg: str):
        self.phase_step = None
        self.phase_total = None
        self._post(msg, "info", "running", True)
    def phase(self, step: int, total: int, msg: str, level: Level = "info"):
        self.phase_step = step
        self.phase_total = total
        self._post(msg, level, "running", False)
    def done(self, msg: str = "완료"):
        self.phase_step = None
        self.phase_total = None
        self._post(msg, "success", "done", False)

File: src/api/progress.ts
import { baseAPI } from "pyloid-js";

export type ProgressLevel = "info" | "success" | "warning" | "error";
export type ProgressStatus = "running" | "done" | "error" | "cancelled" | "unknown";

export type ProgressPayload = {
  step: number;
  total: number;
  phase_step?: number | null;
  phase_total?: number | null;
  level: ProgressLevel;
  status: ProgressStatus;
  message: string;
  error?: string;
  code?: string;
  detail?: string;
  warnings: string[];
  ts: number;
  revision?: number;
};

export const initialProgress: ProgressPayload = {
  step: 0,
  total: 0,
  phase_step: null,
  phase_total: null,
  level: "info",
  status: "unknown",
  message: "",
  error: "",
  detail: "",
  warnings: [],
  ts: 0,
  revision: 0,
};

let rpcEndpoint: string | null = null;
let rpcWindowId: string | null = null;

export async function rpcCallWithTimeout<T>(
  method: string,
  params: Record<string, unknown>,
  timeoutMs: number,
): Promise<T> {
  if (!rpcEndpoint) rpcEndpoint = await baseAPI.getServerUrl();
  if (!rpcWindowId) rpcWindowId = await baseAPI.getWindowId();

  const controller = new AbortController();
  const timeoutId = window.setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(rpcEndpoint, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params, id: rpcWindowId }),
      signal: controller.signal,
    });
    const data = await response.json();
    if (!response.ok) throw new Error(`RPC HTTP ${response.status}`);
    if (data?.error) throw new Error(data.error.message || "RPC 요청에 실패했습니다.");
    return data?.result as T;
  } finally {
    window.clearTimeout(timeoutId);
  }
}

export function normalizeProgress(value: Record<string, unknown>): ProgressPayload {
  return {
    step: Number(value.step ?? 0),
    total: Number(value.total ?? 0),
    phase_step: value.phase_step == null ? null : Number(value.phase_step),
    phase_total: value.phase_total == null ? null : Number(value.phase_total),
    level: (value.level ?? "info") as ProgressLevel,
    status: (value.status ?? "unknown") as ProgressStatus,
    message: String(value.message ?? ""),
    error: value.error == null ? "" : String(value.error),
    code: value.code == null ? "" : String(value.code),
    detail: value.detail == null ? "" : String(value.detail),
    warnings: Array.isArray(value.warnings) ? value.warnings.map(String) : [],
    ts: Number(value.ts ?? Date.now()),
    revision: Number(value.revision ?? 0),
  };
}
